### PR TITLE
Wizard spells

### PR DESCRIPTION
**Pull request by CrudeWax**
Well we are space wizards, aren't we..

### CODE REVIEW

**Comment by LordSaladin** [comment]
el'nath time

File: Resources/Prototypes/Actions/spells.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿- type: action
+  actionType: Fireball
+  icon: Interface/Alerts/Human/human1.png
+  name: "Fireball"
+  filters:
+    - human
+  description: "This is a [color=red]no-wiznerd fireball[/color]."
+  behaviorType: TargetPoint
+  behavior: !type:ProjectileSpell
+   castmessage: EL NATH!
```
Comment (lzimann) [inline Resources/Prototypes/Actions/spells.yml:10]: Cast message for fireball is `ONI SOMA`.
(Plus it's `EI` with an i and not an L)

File: Content.Server/Actions/ProjectileSpell.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        public void ExposeData(ObjectSerializer serializer)
+        {
+            serializer.DataField(this, x => x.CastMessage, "castmessage", "Instant action used."); //What player says upon casting the spell
+            serializer.DataField(this, x => x.Projectile, "spellprojectile", null); //What projectile/Entity does the spell create
+            serializer.DataField(this, x => x.Stationary, "trap", false); //Apply or not apply momentum to the projectile (If true, will spawn a stationary trap-like spell)
+            serializer.DataField(this, x => x.VelocityMult, "speed", 0f); //Speed that is applied to the projectile
+            serializer.DataField(this, x => x.CoolDown, "cooldown", 0f);
+            serializer.DataField(this, x => x.IgnoreCaster, "ignorecaster", false); //ignore caster or not
```
Comment (PaulRitter) [inline Content.Server/Actions/ProjectileSpell.cs:40]: the comments can be xmldocs on the actual properties

File: Content.Server/Actions/ProjectileSpell.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        public ProjectileSpell()
+        {
+            IoCManager.InjectDependencies(this);
+        }
+
+        public void ExposeData(ObjectSerializer serializer)
+        {
+            serializer.DataField(this, x => x.CastMessage, "castmessage", "Instant action used."); //What player says upon casting the spell
+            serializer.DataField(this, x => x.Projectile, "spellprojectile", null); //What projectile/Entity does the spell create
```
Comment (PaulRitter) [inline Content.Server/Actions/ProjectileSpell.cs:36]: add a check below this line to see if the field was set OR give it a proper default value

**Review by PaulRitter — CHANGES_REQUESTED** [review]
looks good, but please seperate the spell for spawning a stationary obj and the spell for spawning a projectile.

right now the projectile gets spawned at the point you clicked, while it should be spawned at the user & be sent to the point you clicked.

File: Content.Server/Actions/ProjectileSpell.cs (hunk 2 above)
Comment (DmitriyRubetskoy) [reply Content.Server/Actions/ProjectileSpell.cs:40]: Please rephrase the suggestion

Comment (metalgearsloth) [reply Content.Server/Actions/ProjectileSpell.cs:40]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/

**Review by PaulRitter — CHANGES_REQUESTED** [review]
you're fireball still gets spawned at the position of your **Click**, not the position of your **entity**.
this is why i suggested seperating the two spells

**Comment by DmitriyRubetskoy** [comment]
I'd like to put a small question to a discussion, and that is concerning how the game should handle the spells, which I consulted mirrorcult and metalgearsloth on, and thankful for their feedback, but still would like to scope and brainstorm what would be the optimal way to handle Touch-range spells.

"Catalyst" or the OG way:
1. Spell is an instant action, and spawns an "activator" item-like entity in the player's hand
2. Activator then has a re-usable component which triggers actual spell effects (written in another component) upon touching someone in item interaction range (the original had activators be the famous "spicy hands", where a player wields a hand-shaped object in their hand to cast the spell). The difference between activators is in their texture and what effect is YML'ed into them, therefore they don't cause much bloat
3. The effect itself is snowflake original code in a separate component, and therefore not reusable 

The two main upsides to it, as I see, is the relative ease of coding it that way, and the intuitive idea of a spell spawning an item in your hand you can use like you would any other item, comforting to both new players and veterans alike. But, if you didn't notice, I am **not a good programmer**, therefore..

The suggested changes:
1. Spell is handled more like an action, by removing the catalyst entity and replacing it with an action input with a entity target (think current disarm action) 
2. The spell telegraphing is shown the same way handcuffs show up on the player, however, there is a TODO on them because there are no inventory overlays coded yet, so someone else would have to code that, until then, the spell would just be shown as a texture layer above the player's avatar
3. The effect is applied upon targeted entity being in range and a disarm-like interaction applying the effect on them, which is again snowflake code component

Any feedback on how to better do this would be welcome.

**Comment by PJB3005** [comment]
The actions system should support selecting targets for spells on it own. "Fake item" should only be reserved for things that *actually require your hand to be free* like pulls. There is nothing "snowflake" about having targeted actions and effects. This is also how SS13 does it and it works fine.

**Comment by DmitriyRubetskoy** [comment]
Yes, but things like cult spell arsenal (mainly stun), are used via creating a "hand" entity imbued with the spell, now, the question is, do we keep it like it was coded in the original (spawn an item into the player's hand, make it so it is free, etc.) or do we do it somehow else?

**Comment by PJB3005** [comment]
> do we keep it like it was coded in the original (spawn an item into the player's hand, make it so it is free, etc.)

With "the original" do you mean SS13? My best recommendation is to stop paying mind to implementation details of whatever crappily coded SS13 server you're looking at.

**Comment by DmitriyRubetskoy** [comment]
image
Example, /tg/ has cult leader with an ability to use "eldrich pulse"

Eldrich pulse, like any other cult spell, creates and entity item in the player's hand (as shown by the ability icon), which upon being used with help intent on a target cultist, will teleport him (the spell effect).

Do we keep the game spawning the "activator" entity in the player's hand (therefore checking if the hand is free, making it easy to telegraph the use due to it being an item texture when held, etc.) or do we use the TargetEntity model and make it so in our game this spell will not check for an open hand, but rather just choose an entity to use it on?

**Comment by PJB3005** [comment]
Is the intent that this item can like, be dropped or handed out to other people?

If not, see my previous comment about shittily coded SS13 servers.

**Comment by metalgearsloth** [comment]
Yeah you should be thinking design-first code-second.

**Comment by DmitriyRubetskoy** [comment]
If it is dropped, it refunds the cooldown and dissapears, that's it. Not something which couldn't be done just by not using the activated action.

I just thought doing so makes it easier to code and add new spells due to it being an item, therefore having the innate ability of using an open hand, telegraphing itself via a held texture, having a visible and examinable physical thing in the inventory, and an inherent touch range. Otherwise, those things would have to be coded somehow differently, and for things like inventory overlays or cursors for actions- those aren't in the game at all.

**Comment by PJB3005** [comment]
> If it is dropped, it refunds the cooldown and dissapears, that's it. Not something which couldn't be done just by not using the activated action.

I hereby declare TG shitcode.

> I just thought doing so makes it easier to code and add new spells due to it being an item

Being a fake physical item always results in *various* funny edge cases that are completely known from SS13 and we should avoid because of hindsight.

> having a visible and examinable physical thing in the inventory

Having to examine to see what spell a wizard is channeling and probably about to throw at your face is atrocious. Don't even consider it.

> and an inherent touch range

Even this is way more fragile than you're probably imagining it to be. /vg/ has stuff like borer arms which allows you to extend your hand reach, etc... If you need shared functionality of interaction distances there are better ways to achieve that.

> Otherwise, those things would have to be coded somehow differently, and for things like inventory overlays or cursors for actions- those aren't in the game at all.

Then code them.

**Comment by DmitriyRubetskoy** [comment]
How do you address telegraphing and taking up hands?

**Comment by DmitriyRubetskoy** [comment]
The only viable option for telegraphing I see is that upon pressing the button the "spicy hand" appears, and only dissapears after the use of the spell, therefore all the time it is "prepared" to be cast it lights up, but I am unsure if that is a good idea

**Comment by PJB3005** [comment]
Taking up hands should be explicitly supported by the hands component, probably. Although what spells even need a free hand? 

For telegraphing just code a visualizer or something? I don't see the problem here.

**Comment by DmitriyRubetskoy** [comment]
At what moment do you telegraph the spell which does not have a physical manifestation?
Upon cast? Upon the ability being activated?

**Comment by DmitriyRubetskoy** [comment]
With items it is simple "As long as the player holds the activator" but how do you do it otherwise?
Also are you suggesting most original spells to be made usable without an open hand?

**Comment by PJB3005** [comment]
> Upon cast? Upon the ability being activated?

I would say when you activate the spell in the action bar to select a target. If the player cancels the spell (doesn't select a target) it just... stops right there.

> Also are you suggesting most original spells to be made usable without an open hand?

Half the spells already work like this on /vg/ and the other spells seem like the "can't interact with computer while holding pen" all over, so yes.

**Comment by DmitriyRubetskoy** [comment]
Issue resolved.
We make it so the spells aren't telegraphed at all (yet), and for now I will just focus on adding the basic cogs and gears of a target entity spell

**Comment by DmitriyRubetskoy** [comment]
Alright, this shit has too much interlocking and confusing parts?
How about I close this draft, and PR each completed spell type seperately, along with a debug spellbook to test them. Sounds good?

**Review by PaulRitter — CHANGES_REQUESTED** [review]
few notes:
- update to use serv3
- move comments in the exposedata to individual xmldocs on the fields
- dont use "Instant action used." as a castmessage default. just leave it empty & not send one if not mapped IMO
- make sure to merge latest master to get your yaml/rsis validated

**Comment by DmitriyRubetskoy** [comment]
1. Commit takes this long due to me being in the middle of reworking for serv3
2. Will be done, these are just for personal naviagtion
3. Sure

**Comment by SweptWasTaken** [comment]
Stale. Closing for now.

**Comment by DmitriyRubetskoy** [comment]
Still working on it, just having problems re-writing chunks of code for adapting to serv3

**Comment by DmitriyRubetskoy** [comment]
Alright, what do we have here?

3 spell types: 
- Projectilespell.cs, creates an entity and applies velocity to it, if so desired. (Fireball, Traps!, Summon Carp)
- Targetspell.cs or touch spell, applies a component, if the target within a certain distance has another component, also has a MagicalProjectile.cs variant applicable to projectile entities (BarnYard curse [projectile], Pacify [touch-spell])
- ItemSpell.cs gives an item to the caster (summon pie)

All of these contain customizable datafields, such as cooldowns, spell messages, entities used, sound when cast, etc.
Spell projectiles are handled by wizard summons yaml which stores spell-related projectiles and entities. 

TODOS via seperate commits:
-Spell serialisation. Component spells could be expanded to have tweakable values in applied components.
-Spell shouts being actual speech
-"Cursed" alert and status effect

**Comment by DmitriyRubetskoy** [comment]
image
The error output on build has nothing to do with my commit so I am not sure what I can do to aid that

**Comment by metalgearsloth** [comment]
> 
> 
> image
> The error output on build has nothing to do with my commit so I am not sure what I can do to aid that

Your PR adds it back in

**Comment by metalgearsloth** [comment]
Fix linter if you want a full review.

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Ammunition/explosives.yml [hunk 4]
```diff
@@ -13,6 +13,22 @@
     sprite: Objects/Weapons/Guns/Ammunition/Explosives/rpg.rsi
     state: frag
     drawdepth: FloorObjects
+    
+- type: entity
+  id: FireBall
+  name: PlaceHolder fire
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Objects/Weapons/Guns/Ammunition/explosives.yml:19]: ?

File: Content.Server/Actions/Spells/ProjectileSpell.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            IoCManager.InjectDependencies(this);
+        }
+
+        public void DoTargetPointAction(TargetPointActionEventArgs args)
+        {
+            var caster = args.Performer;
+            if (!caster.TryGetComponent<SharedActionsComponent>(out var actions)) return;
+            actions.Cooldown(args.ActionType, Cooldowns.SecondsFromNow(CoolDown)); //Set the spell on cooldown
+            var playerPosition = args.Performer.Transform.LocalPosition; //Set relative position of the entity of the spell (used later)
+            var direction = (args.Target.Position - playerPosition).Normalized * 2; //Decides the general direction of the spell (used later) + how far it goes
```
Comment (metalgearsloth) [inline Content.Server/Actions/Spells/ProjectileSpell.cs:45]: Same with this, also the vector could be 0 which can't be normalized, also why is it being multiplied by 2 and used as the velocity?

File: Content.Server/Actions/Spells/InstantSpell.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        //Rubber-band snapping items into player's hands, originally was a workaround, later found it works quite well with stuns
+        //Not sure if needs fixing
+
+        public void DoInstantAction(InstantActionEventArgs args)
+        {
+            if (!args.Performer.TryGetComponent<SharedActionsComponent>(out var actions)) return;
+            var caster = args.Performer;
+            var casterCoords = caster.Transform.Coordinates;
+            var spawnedProto = caster.EntityManager.SpawnEntity(ItemProto, casterCoords);
+            caster.PopupMessageEveryone(CastMessage);
```
Comment (metalgearsloth) [inline Content.Server/Actions/Spells/InstantSpell.cs:38]: Rest of the popups only popup for caster which isn't consistent.

File: Content.Server/Actions/Spells/ProjectileSpell.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            {
+                if (IgnoreCaster == true) //ignores or not the caster
+                {
+                    projectileComponent.IgnoreEntity(args.Performer);
+                }
+                spawnedSpell
+                .GetComponent<PhysicsComponent>()
+                .LinearVelocity = direction * VelocityMult;
+            }
+              spawnedSpell.Transform.LocalRotation = args.Performer.Transform.LocalRotation;
```
Comment (metalgearsloth) [inline Content.Server/Actions/Spells/ProjectileSpell.cs:62]: Should be WorldRotation?

Also a bit messy to duplicate code to spawn a projectile and also spawn a whatever. Same with taking in IgnoreCaster when the sole purpose of projectilecomp is to allow a specific entity to be ignored. You're also duplicating GetComponent.

File: Content.Server/Actions/Spells/ProjectileSpell.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            var caster = args.Performer;
+            if (!caster.TryGetComponent<SharedActionsComponent>(out var actions)) return;
+            actions.Cooldown(args.ActionType, Cooldowns.SecondsFromNow(CoolDown)); //Set the spell on cooldown
+            var playerPosition = args.Performer.Transform.LocalPosition; //Set relative position of the entity of the spell (used later)
+            var direction = (args.Target.Position - playerPosition).Normalized * 2; //Decides the general direction of the spell (used later) + how far it goes
+            var coords = args.Performer.Transform.Coordinates.WithPosition(playerPosition + direction);
+
+            caster.PopupMessageEveryone(CastMessage); //Speak the cast message out loud
+
+            var spawnedSpell = _entityManager.SpawnEntity(Projectile, coords);
```
Comment (metalgearsloth) [inline Content.Server/Actions/Spells/ProjectileSpell.cs:50]: Should probably be spawned on mappos.

File: Content.Server/Actions/Spells/TargetSpell.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            if (!target.TryGetComponent(RegisteredTargetType, out var component))
+            {
+                caster.PopupMessage("Your hands fizzle, your target is invalid!");
+                return;
+            }
+            if (target.HasComponent(RegisteredInduceType))
+            {
+                caster.PopupMessage("This poor soul is already cursed!");
+                return;
+            }
```
Comment (metalgearsloth) [inline Content.Server/Actions/Spells/TargetSpell.cs:69]: If using these they all need to be localised.

File: Content.Server/Actions/Spells/TargetSpell.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            //For the range of the spell
+            var casterCoords = caster.Transform.WorldPosition;
+            var targetCoords = target.Transform.WorldPosition;
+            var effectiveRange = (casterCoords - targetCoords).Length;
+            if (!caster.TryGetComponent<SharedActionsComponent>(out var actions)) return;
+            if (CastRange < effectiveRange && CastRange != 0)
+            {
+                caster.PopupMessage("This spell cannot reach this far!");
+                return;
+            }
```
Comment (metalgearsloth) [inline Content.Server/Actions/Spells/TargetSpell.cs:57]: You should always check cheap stuff first (castrange)

File: Resources/Prototypes/Actions/spells.yml [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    - spells
+  description: "Open a portal into space dragon's lair to summon a carp imbued with magic!"
+  behaviorType: TargetPoint
+  behavior: !type:ProjectileSpell
+   castmessage: RI'FATA!
+   spellprojectile:  MagicarpMob_Content
+   range: 2
+   speed: 0
+   cooldown: 1
+   ignorecaster: true
```
Comment (metalgearsloth) [inline Resources/Prototypes/Actions/spells.yml:45]: As above re-using "ProjectileSpell" for this is messy. Same with the trap.

File: Resources/Prototypes/Entities/Objects/Specific/Magical/SpellBooks.yml [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+  name: Urban Legends
+  parent: BaseItem
+  id: Debug Spellbook
+  description: 'penned by "Beyond Occult Rituals Typography"'
+  components:
+  - type: Sprite
+    sprite: Objects/Misc/books.rsi
+    state: bookold
+  - type: SpellBook
+    usemessage: You're a wiznred, Garry!
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Objects/Specific/Magical/SpellBooks.yml:11]: 1 example of the spelling mistakes.

**Comment by DmitriyRubetskoy** [comment]
Review received, I'll get to it

File: Resources/Prototypes/Entities/Objects/Specific/Magical/SpellBooks.yml (hunk 12 above)
Comment (DmitriyRubetskoy) [reply Resources/Prototypes/Entities/Objects/Specific/Magical/SpellBooks.yml:11]: oh that's actually just a "You're are a wizard, Harry!" but jumbled

File: Resources/Prototypes/Actions/spells.yml (hunk 11 above)
Comment (DmitriyRubetskoy) [reply Resources/Prototypes/Actions/spells.yml:45]: Why is it messy? In gameplay it works fine and it saves having to bloat the game with "summon this, but stationary" script

File: Resources/Textures/Objects/Weapons/Guns/Projectiles/magicboltfizzy.rsi/meta.json [hunk 13]
```diff
@@ -0,0 +1,20 @@
+{
+  "version": 1,
+  "license": "CC-BY-SA-3.0",
+  "copyright": "Original asset provided by https://github.com/DmitriyRubetskoy",
```
[the remark this answers is not in the document]
Comment (DmitriyRubetskoy) [reply Resources/Textures/Objects/Weapons/Guns/Projectiles/magicboltfizzy.rsi/meta.json:4]: Nope, I made it from scratch

File: Content.Server/Actions/Spells/SpellComponents/PacifyComponent.cs [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        bool IActionBlocker.CanAttack() => false;
+
+        bool IActionBlocker.CanEquip() => true;
+
+        bool IActionBlocker.CanUnequip() => true;
+
+
+        #endregion
+
+    }
```
[the remark this answers is not in the document]
Comment (DmitriyRubetskoy) [reply Content.Server/Actions/Spells/SpellComponents/PacifyComponent.cs:45]: Nope, it is used by a related spell. It's a really basic demo of a spell component.

File: Content.Server/GameObjects/Components/SpellBookComponent.cs [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            if (!eventArgs.User.TryGetComponent<SharedActionsComponent>(out var actions)) return false;
+            if (GrantedSpells == null) return false;
+            foreach(var spell in GrantedSpells)
+            {
+                actions.Grant(spell);
+            }
+            eventArgs.User.PopupMessage(UseMessage);
+            Owner.Delete();
+            return true;
```
[the remark this answers is not in the document]
Comment (DmitriyRubetskoy) [reply Content.Server/GameObjects/Components/SpellBookComponent.cs:34]: I assumed that'd be handled in the wizard antag commit

File: Content.Server/GameObjects/Components/Projectiles/MagicalProjectileComponent.cs [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+            Component compInducedFinal = (Component) componentInduced;
+            compInducedFinal.Owner = target;
+            target.EntityManager.ComponentManager.AddComponent(target, compInducedFinal);
+            target.SpawnTimer(SpellDuration, () => target.EntityManager.ComponentManager.RemoveComponent(target.Uid, compInducedFinal));
+            if (_castsound != null)
+            {
+                SoundSystem.Play(Filter.Pvs(Owner), _castsound, Owner);
+            }
+            else return;
+        }
```
[the remark this answers is not in the document]
Comment (DmitriyRubetskoy) [reply Content.Server/GameObjects/Components/Projectiles/MagicalProjectileComponent.cs:59]: Will do

File: Content.Server/Actions/Spells/ProjectileSpell.cs (hunk 5 above)
Comment (DmitriyRubetskoy) [reply Content.Server/Actions/Spells/ProjectileSpell.cs:41]: I don't particularly remember to be honest, I distinctly remember it being somehow more accurate (?)

File: Content.Server/Actions/Spells/InstantSpell.cs (hunk 6 above)
Comment (DmitriyRubetskoy) [reply Content.Server/Actions/Spells/InstantSpell.cs:34]: Those which don't popup for everyone are utility-related and doesn't need to be seen.
such as fails to cast, to reach, etc

File: Resources/Prototypes/Entities/Objects/Specific/Magical/SpellBooks.yml (hunk 12 above)
Comment (ShadowCommander) [reply Resources/Prototypes/Entities/Objects/Specific/Magical/SpellBooks.yml:11]: Maybe change it to wiznerd?

Comment (DmitriyRubetskoy) [reply Resources/Prototypes/Entities/Objects/Specific/Magical/SpellBooks.yml:11]: Oh sorry, my bad

File: Content.Server/GameObjects/Components/Projectiles/MagicalProjectileComponent.cs [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+using Robust.Shared.Player;
+using Robust.Shared.Serialization;
+using Robust.Shared.Serialization.Manager.Attributes;
+using Robust.Shared.ViewVariables;
+using System;
+
+namespace Content.Server.GameObjects.Components.Projectiles
+{
+    [RegisterComponent]
+    public class MagicalProjectileComponent : Component, IStartCollide
```
[the remark this answers is not in the document]
Comment (DmitriyRubetskoy) [reply Content.Server/GameObjects/Components/Projectiles/MagicalProjectileComponent.cs:14]: Noted

File: Resources/Prototypes/Entities/Objects/Specific/Magical/SpellBooks.yml [hunk 18]
```diff
@@ -0,0 +1,18 @@
+- type: entity
+  name: Urban Legends
+  parent: BaseItem
+  id: Debug Spellbook
+  description: 'penned by "Beyond Occult Rituals Typography"'
```
[the remark this answers is not in the document]
Comment (DmitriyRubetskoy) [reply Resources/Prototypes/Entities/Objects/Specific/Magical/SpellBooks.yml:5]: It is also made like this in order to 
A. Make a showoff for single-use spellbooks in general
B. Adminbus option maybe, if someone wants to make a fun round without wizard being in the game as a full-on antagonist yet

File: Content.Server/GameObjects/Components/Projectiles/MagicalProjectileComponent.cs [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        public override string Name => "MagicalProjectile";
+
+        [ViewVariables] [DataField("NeedComponent")] public string TargetType { get; set; } = "Mind";
+
+        [ViewVariables] [DataField("AddedComponent")] public string InduceComponent { get; set; } = "RadiatonPulse";
+
+        [ViewVariables] [DataField("duration")] public int SpellDuration { get; set; } = 100;
+
+        [ViewVariables] [DataField("castsound")] private string? _castsound = "/Audio/Weapons/emitter.ogg";
```
[the remark this answers is not in the document]
Comment (DmitriyRubetskoy) [reply Content.Server/GameObjects/Components/Projectiles/MagicalProjectileComponent.cs:24]: Will be turned into a more generic component

File: Content.Server/Actions/Spells/ProjectileSpell.cs (hunk 8 above)
Comment (DmitriyRubetskoy) [reply Content.Server/Actions/Spells/ProjectileSpell.cs:46]: Noted

File: Resources/Prototypes/Actions/spells.yml [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+   cooldown: 1
+   ignoreCaster: true
+
+- type: action
+  actionType: SpellTraps
+  icon: Interface/Actions/Spells/trap.png
+  name: "Traps!"
+  filters:
+    - spells
+  description: "Snare your enemies with a magical trap! Semi-transperent."
```
Comment (Mith-randalf) [inline Resources/Prototypes/Actions/spells.yml:22]: lil typo here
*Semi-transparent

Comment (DmitriyRubetskoy) [reply Resources/Prototypes/Actions/spells.yml:22]: Thanks for the attention

**Comment by metalgearsloth** [comment]
Are you still working on this crude?

**Comment by SweptWasTaken** [comment]
> 
> 
> Are you still working on this crude?

Assuming not. Closing.

**Comment by DmitriyRubetskoy** [comment]
Still working on it. Taking a break due to life situation. I'll reopen it once I fix it.

**Comment by DmitriyRubetskoy** [comment]
I will split the PR into smaller ones for a faster merge

**Pull request closed without merging**